### ISSUE TITLE
Support identifying potentially errorneous characters

### ISSUE THREAD

**Issue by Kixunil** [rust-bitcoin/rust-bech32#95]
The checksum algorithm has error detection that can identify which characters were actually wrong (with some probability). It'd be great to support this in the library.

**Comment by clarkmoody** [comment]
From the text of BIP-0173:

> Because of this, implementations SHOULD NOT implement correction beyond potentially suggesting to the user where in the string an error might be found, without suggesting the correction to make.

I would support including some position information in one of the enum variants. This feature might necessitate a new API function that decodes with detection, since it seems to be a more intense operation than simple checksum validation.

**Comment by Kixunil** [comment]
Indeed, that's why I wrote "identify", not "correct" :)

And yes, position information should be in error variant. I wonder if feature flag is enough.

**Comment by Kixunil** [comment]
OK, that makes sense for this library. But what about `bitcoin::Address`? It'd suck to not have locations returned from `from_str`. I guess I'd skip them in case of `serde` deserialization though.

**Comment by apoelstra** [comment]
For `bitcoin::Address` we probably also don't want a 50x slowdown in parsing.

Though I wouldn't mind if `Address::from_str`did this and we had a separate `parse_without_error_locations` method or something that performance-minded people could use. And serde sholud not do locating, agreed.

**Comment by Kixunil** [comment]
Yeah, also we would only calculate them on errors, which should be cold path in vast majority of applications.

**Comment by apoelstra** [comment]
Yeah, that's a good point -- the first step of error locating is doing the normal verification anyway, and we know immediately whether there is even any point in doing the heavy computations.

**Comment by apoelstra** [comment]
I'm going to close this. The issue seems to be "we should implement error correction" which is mostly working and does not need a tracking issue. (Or if it does, it is not this issue.)

**Comment by Kixunil** [comment]
This is not about error **correction**. Are you sure you intended to close this?

**Comment by apoelstra** [comment]
"error detection that can identify which characters were actually wrong" is error correction.

**Comment by Kixunil** [comment]
I don't think so? Correction would be identifying which characters were supposed to be there, which we explicitly should NOT do.

**Comment by apoelstra** [comment]
You are welcome to open an issue describing exactly what functionality you want (and on what objects).

The current OP asks for "error detection that can identify which characters were actually wrong" which is implemented in the library. (It also mentions probabilities which we cannot do because we have no information about the distribution of errors that we might be seeing, though a more advanced algorithm could use such information to produce a distribution of corrections.)

Remember that this library is used for more than the bech32 checksum, and the bech32 checksum is used for more than just Bitcoin addresses, so it is unlikely that there will be functionality that we should universally "should" or "should NOT" do.
